types/src/primary.rs

```diff
@@ -381,6 +386,48 @@ impl Certificate {
             .collect()
     }
 
+    pub fn new(
+        committee: &Committee,
+        header: Header,
+        votes: Vec<(PublicKey, Signature)>,
+    ) -> DagResult<Certificate> {
+        let mut votes = votes;
+        votes.sort_by_key(|(pk, _)| pk.clone());
+
+        let mut votes_iter = 0;
+        let votes_len = votes.len();
+        let keys = committee.keys();
+        let filtered_votes = keys
+            .iter()
+            .enumerate()
+            .filter(|(_, &pk)| {
+                if votes_iter < votes_len && pk == &votes[votes_iter].0 {
+                    votes_iter += 1;
+                    return true;
+                }
+                false
+            })
+            .map(|(index, _)| index as u32);
+        let signed_authorities= roaring::RoaringBitmap::from_sorted_iter(filtered_votes)
+            .map_err(|_| DagError::InvalidBitmap("Failed to convert votes into a bitmap of authority keys. Something is likely very wrong...".to_string()))?;
+
+        // Ensure that all authorities in the set of votes are known
+        ensure!(
+            votes_iter == votes_len,
+            DagError::UnknownAuthority(votes[votes_iter].0.encode_base64())
+        );
+
+        let aggregated_signature =
+            AggregateSignature::aggregate(votes.iter().map(|(_, sig)| sig).cloned().collect())
+                .map_err(DagError::InvalidSignature)?;
+
+        Ok(Certificate {
+            header,
+            aggregated_signature,
+            signed_authorities,
+        })
+    }
```

Comment:
We will need tests on this. More specifically:
- that a set of votes using a `PublicKey` twice results in an error,
- unless it's providing twice the exact same set of bytes for the signatures in which case we should just log a warning and include the signature once,
- that a set of votes with an aggregated stake lower than the `committee` threshold results in an error, (same logic as in validation)
- please also insert a randomized test (you can use [proptest](https://docs.rs/proptest/latest/proptest/)) showing that given an arbitrary committee, and an arbitrary set of valid votes by 50% of it (such that 50% passes the threshold stake), the output of `Certificate::new` passes `verify()`.